src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value_impl.hpp

```diff
@@ -0,0 +1,457 @@
+/**
+ * @file discrete_hilbert_value.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Defintion of the DiscreteHilbertValue class, a class that calculates
+ * the ordering of points using the Hilbert curve.
+ */
+#ifndef MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_IMPL_HPP
+#define MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_IMPL_HPP
+
+#include "discrete_hilbert_value.hpp"
+
+namespace mlpack {
+namespace tree /** Trees and tree-building procedures. */ {
+
+template<typename TreeElemType>
+DiscreteHilbertValue<TreeElemType>::DiscreteHilbertValue() :
+    localDataset(NULL),
+    ownsLocalDataset(false),
+    numValues(0),
+    valueToInsert(NULL),
+    ownsValueToInsert(false)
+{
+
+}
+
+template<typename TreeElemType>
+DiscreteHilbertValue<TreeElemType>::~DiscreteHilbertValue()
+{
+  if (ownsLocalDataset)
+    delete localDataset;
+  if (ownsValueToInsert)
+    delete valueToInsert;
+}
+
+template<typename TreeElemType>
+template<typename TreeType>
+DiscreteHilbertValue<TreeElemType>::DiscreteHilbertValue(const TreeType* tree) :
+    localDataset(NULL),
+    ownsLocalDataset(false),
+    numValues(0),
+    valueToInsert(tree->Parent() ? 
+                tree->Parent()->AuxiliaryInfo().HilbertValue().ValueToInsert() :
+                new arma::Col<HilbertElemType>(tree->Dataset().n_rows)),
+    ownsValueToInsert(tree->Parent() ? false : true)
+{
+  // Calculate the Hilbert value for all points
+  if (!tree->Parent()) //  This is the root node
+    ownsLocalDataset = true;
+  else if (tree->Parent()->Children()[0]->IsLeaf())
+  {
+    // This is a leaf node
+    assert(tree->Parent()->NumChildren() > 0);
+    ownsLocalDataset = true;
+  }
```

Comment:
Okay, I understand.  I see that each node holds a `DiscreteHilbertValue` object, and each `DiscreteHilbertValue` object stores its own local vector of Hilbert values for each point contained in the node.  So it makes sense to keep the `localDataset` object so we can avoid resizing the Hilbert values vector, but I think we should rename it to `localHilbertValues` or something like that.